hoss/dimension_utilities.py

```diff
@@ -101,10 +101,214 @@ def prefetch_dimension_variables(
 
     # Create bounds variables if necessary.
     add_bounds_variables(required_dimensions_nc4, required_dimensions, varinfo, logger)
-
     return required_dimensions_nc4
 
 
+def get_override_projected_dimensions(
+    varinfo: VarInfoFromDmr,
+    override_variable_name: str,
+) -> str | None:
+    """returns the x-y projection variable names that would
+    match the geo coordinate names. The `latitude` coordinate
+    variable name gets converted to 'projected_y' dimension scale
+    and the `longitude` coordinate variable name gets converted to
+    'projected_x'
+
+    """
+    override_variable = varinfo.get_variable(override_variable_name)
+    if override_variable is not None:
+        if override_variable.is_latitude():
+            projected_dimension_name = 'projected_y'
+        elif override_variable.is_longitude():
+            projected_dimension_name = 'projected_x'
+        else:
+            projected_dimension_name = None
+    return projected_dimension_name
+
+
+def get_coordinate_variables(
+    varinfo: VarInfoFromDmr,
+    requested_variables: Set[str],
+) -> list[str]:
+    """This method returns coordinate variables that are referenced
+    in the variables requested. It returns it in a specific order
+    [latitude, longitude]
+    """
+
+    # try:
+    coordinate_variables_set = varinfo.get_references_for_attribute(
+        requested_variables, 'coordinates'
+    )
+    coordinate_variables = []
+    for coordinate in coordinate_variables_set:
+        if varinfo.get_variable(coordinate).is_latitude():
+            coordinate_variables.insert(0, coordinate)
+        elif varinfo.get_variable(coordinate).is_longitude():
+            coordinate_variables.insert(1, coordinate)
+
+    return coordinate_variables
+    # except AttributeError:
+    #    return set()
+
+
+def update_dimension_variables(
+    prefetch_dataset: Dataset,
+    coordinates: Set[str],
+    varinfo: VarInfoFromDmr,
+) -> Dict[str, ndarray]:
+    """Generate artificial 1D dimensions variable for each
+    2D dimension or coordinate variable
+
+    For each dimension variable:
+    (1) Check if the dimension variable is 1D.
+    (2) If it is not 1D and is 2D get the dimension sizes
+    (3) Get the corner points from the coordinate variables
+    (4) Get the x-y max-min values
+    (5) Generate the x-y dimscale array and return to the calling method
+
+    """
+    for coordinate_name in coordinates:
+        coordinate_variable = varinfo.get_variable(coordinate_name)
+        if prefetch_dataset[coordinate_variable.full_name_path][:].ndim > 1:
+            col_size = prefetch_dataset[coordinate_variable.full_name_path][:].shape[0]
+            row_size = prefetch_dataset[coordinate_variable.full_name_path][:].shape[1]
```

Comment:
Two thoughts here:

1) You can access the shape of the array directly from the variable object, e.g.:

```
col_size = prefetch_dataset[coordinate_variable.full_name_path].shape[0]
row_size = prefetch_dataset[coordinate_variable.full_name_path].shape[1]
```

(It seems like a little change, but I would guess the way that the `netCDF4.Variable` class works means that you are loading a lot less information into memory until you need it)

2) This is another instance of a problem with variable declarations happening inside a conditional block, then those variables being used outside the scope of the conditional. If the number of dimensions on the variable is 0 or 1, then you are going to skip the declaration of `col_size` and `row_size`. But then those variables will still be used in the calculation of the resolutions. This suggests that you don't either need to provide alternative values in an `else`, or that you don't have enough in the conditional statement. (Thorough unit tests would really help with these sorts of issues)

3) This function is likely part of the problem with 3-D variables. You are making an assumption that the first two axes of the array are relating to spatial information. Given how Python indices work, it's most likely the last two (e.g., `(time, lat, lon)`), but that ordering should not be counted on. Instead of assuming ordering of the array axes, how could you know for sure?

Reply:
1) I think the col_size and row_size, I had changed it already with the shape
2) I think I refactored the method in one of the commits.
3) for ordering - in the case of coordinates - I can check for lat and lon and maintain the order that I need. will check how it works for a 3D variable example. I think I have an example to test

Reply:
the 3D cases do not work for SPL3SMP - the coordinates are 2D.
This works for opendap direct url
landcover_class[ 0..405] [ 0..963] [ 0..2] is the grid 
9:38  487:595  0:2 works on opendap

The hoss code has 
variables_with_ranges: /Soil_Moisture_Retrieval_Data_AM/landcover_class[9:38][487:595]
 
 I can update the test case and results in the ticket DAS-2238